### ISSUE TITLE
cleaning up cs variants

### ISSUE THREAD

**Issue by FrankMittelbach** [latex3/latex3#1257]
We do have by now a good number of variants accumulated via `\cs_generate_variant:Nn` in our own code and in addition there is a good number added in external packages so it is perhaps time to take stock and move some of them to the l3kernel and then remove them elsewhere. 

The overview shows the usage by package (taking out the variants that are clearly package only):

[variants-overview.md](https://github.com/latex3/latex3/files/12266824/variants-overview.md)

And this is a slightly cleaned version sorted by command and occurances. The numbers aren't right because sometimes people declared several variants in one go and they are then counted as a separate case.

[variants-sorted.md](https://github.com/latex3/latex3/files/12266828/variants-sorted.md)

Some usages seem a bit odd on first glance  and things should be done differently perhaps.

And the fact a variant is declared somewhere doesn't mean that it is not already part of the kernel, so the list would need some further cleanup for sure.

Certainly low priority, but it does make sense that we do move used variants to the kernel, so that they show up in `interface3` as available in the future. At least for those that are declared multiple times I think we should do this I guess and certainly for all that are used inside the 2e kernel or its core support packages.

**Comment by josephwright** [comment]
Certainly worth picking off common cases, at least where they make sense

I notice quite a few `V`-type variants for `int` functions, for example, that are really not needed at all. So there is also some user education to do.

**Comment by josephwright** [comment]
BTW, worth noting that adding variants to the kernel doesn't immediately mean that the generation can be removed from packages: https://github.com/josephwright/siunitx/issues/680 :)

**Comment by FrankMittelbach** [comment]
@josephwright you mean because people only update packages (manually) but not LaTeX and the L3 layer of LaTeX? Yes I guess updating anything outside of the kernel (and the core packages that go with it) is best done only after a full major release of TL and even then you may have to set \NeedsFormat with a date to be on the safe side.

**Comment by FrankMittelbach** [comment]
> I notice quite a few `V`-type variants for `int` functions, for example, that are really not needed at all. So there is also some user education to do.

people might like them nevertheless for reasons of consistent look and feel. So perhaps they should be provided but hand-tuned by the kernel (to hide the internal reason for them not being necessary).

**Comment by josephwright** [comment]
> > I notice quite a few `V`-type variants for `int` functions, for example, that are really not needed at all. So there is also some user education to do.
> 
> people might like them nevertheless for reasons of consistent look and feel. So perhaps they should be provided but hand-tuned by the kernel (to hide the internal reason for them not being necessary).

Actually, I realise there is an end-case where this does make sense: if you want to pass the content of a toks to a numerical expression:
```
\ExplSyntaxOn
\cs_generate_variant:Nn \int_set:Nn { NV }
\ExplSyntaxOff

\makeatletter
\toks@={100}\relax
\ExplSyntaxOn
\int_zero:N \l_tmpa_int
\int_set:Nn \l_tmpa_int \toks@
\int_show:N \l_tmpa_int
```

**Comment by josephwright** [comment]
> @josephwright you mean because people only update packages (manually) but not LaTeX and the L3 layer of LaTeX? 

Yes, that's certainly how it looks for that `siunitx` issue report. (Aside: do you think I should be running tests for that? The only way I could think of is running my entire test suite with a parallel config that uses the oldest `expl3` release I say I support. Is this sensible/best practice?)

**Comment by josephwright** [comment]
At least for all of the one-argument functions, I suspect we should just do 'the lot' (all variants). Similarly, for functions with two arguments where one is `N`-type, we could provide everything as it's a full set only for one argument, and only two variants (`N`/`c`) for the second. Thoughts?

**Comment by muzimuzhi** [comment]
> At least for all of the one-argument functions, I suspect we should just do 'the lot' (all variants). Similarly, for functions with two arguments where one is `N`-type, we could provide everything as it's a full set only for one argument, and only two variants (`N`/`c`) for the second. Thoughts?

#1258 need to be resolved first (or at the same time), otherwise the poor output described in #1258 may appear more frequently.

**Comment by car222222** [comment]
I agree that #1258 needs some resolution RSN!

**Comment by FrankMittelbach** [comment]
> Actually, I realise there is an end-case where this does make sense: if you want to pass the content of a toks to a numerical expression:

well, yes toks are a sort of use case. On the other hand toks are deliberately not part of the L3 layer (except very deep down, so I think it is a bit on the questionable side.

**Comment by FrankMittelbach** [comment]
> > @josephwright you mean because people only update packages (manually) but not LaTeX and the L3 layer of LaTeX?
> 
> Yes, that's certainly how it looks for that `siunitx` issue report. (Aside: do you think I should be running tests for that? The only way I could think of is running my entire test suite with a parallel config that uses the oldest `expl3` release I say I support. Is this sensible/best practice?)

perhaps to perfect, but it might already help if you add a few tests that checks if a current siunitx loads (and executes a few things when your test puts `\RequirePackage[date]{latexrelease}` on top, so having a few such tests might be already going a long way (dunno if true)

**Comment by FrankMittelbach** [comment]
> > 
> 
> perhaps to perfect, but it might already help if you add a few tests that checks if a current siunitx loads (and executes a few things when your test puts `\RequirePackage[date]{latexrelease}` on top, so having a few such tests might be already going a long way (dunno if true)

On the other hand, maybe not, at least not for the cases where the issue is that you need to have a recent L3 prog layer, because that is not rolled back. However, if you require L3 prog layer from date X then probably you should require that the format is of a date that this prog layer is included Or perhaps even test for the prog layer explicitly (as suggested in one of the recent enhancement requests).

**Comment by josephwright** [comment]
With https://github.com/latex3/latex3/issues/1258 fixed, I guess I can start picking stuff off here: do people like my plan for the one and one-plus-N cases?

**Comment by FrankMittelbach** [comment]
> At least for all of the one-argument functions, I suspect we should just do 'the lot' (all variants). Similarly, for functions with two arguments where one is `N`-type, we could provide everything as it's a full set only for one argument, and only two variants (`N`/`c`) for the second. Thoughts?

I think there is nothing wrong with that but I would do it only for those where we do have evidence that there is at least one or two uses outside of the base function (ie starting from my list), I mean we are still blowing token space and csnames in the format and if they aren't ever needed then ...

**Comment by davidcarlisle** [comment]
> I suspect we should just do 'the lot' (all variants). 

We could consider not including `x`  in the variants that are added "just in case" (and prefer `e`) just include `x`  known to be used (ie in Frank's list).  

Also any expandable functions that have non expandable x variants that don't have an e variant should probably have the e version added.

**Comment by josephwright** [comment]
> > I suspect we should just do 'the lot' (all variants).
> 
> We could consider not including `x` in the variants that are added "just in case" (and prefer `e`) just include `x` known to be used (ie in Frank's list).
> 
> Also any expandable functions that have non expandable x variants that don't have an e variant should probably have the e version added.

I'll check first over this before starting on the main list: hopefully I can look at this over the coming week and do a tidy-up release.

**Comment by car222222** [comment]
Good to know that at least someone is thinking about this: 
blowing token space and csnames !

**Comment by josephwright** [comment]
@car222222  You might remember that we talked before about 'just adding all variants' and ruled it out for that reason!

**Comment by car222222** [comment]
@josephwright a long time before??  I must have missed that decision ??

**Comment by Skillmon** [comment]
@car222222 in this thread, not too long ago, there was this: https://github.com/latex3/latex3/issues/1257#issuecomment-1675005611

**Comment by josephwright** [comment]
> @josephwright a long time before?? I must have missed that decision ??

Goes back years: we did take about rounding out variants on the team list at least a decade ago. For example, the reason that the `msg` functions are only `:nnnn` and `:xxxx` was to avoid an explosion, as you can always `\exp_not:<thing> { ... }` in the all-`x` versions.

**Comment by car222222** [comment]
@Skillmon Yes, I referred to that already.

**Comment by josephwright** [comment]
Taking my list and removing those entries that already exist, for cases of 2 or more uses of the variant I have 147 potential additions. Some of these are ones I'm not keen on (e.g. `x`-type for expandable functions), but I think we can likely justify most of them. I will look at the 'created once only'  set separately: there are a lot more and it will be quite some work to grind through them.

**Comment by car222222** [comment]
147 is very reasonable.  
Is it also sensible to try and eradicate the use of x-variants of otherwise fully expandable functions?  By, at least, contacting their current users, where possible, and then deprecating them.

**Comment by josephwright** [comment]
> it also sensible to try and eradicate the use of x-variants of otherwise fully expandable functions? By, at least, contacting their current users, where possible, and then deprecating them.

Fits in with my plan. I am excluding any `x`-type variants for expandable functions, and will check we have a suitable `e`-type in a second pass. We can then look at asking for changes.

I'm also planning not to add `f`-type variants in general, as they can almost certainly be covered by `e`-type. There are perhaps a couple of cases where space treatment might be important: I'll look hard at those as  I work through.

In a similar area, for non-expandable functions we've mainly got `x`-type, and I'm sticking with that. (The cases where we have `e`-type in the kernel here are mainly to support `lthooks`, where the difference in treatment of `#` chars is important.)

---

My plan is to do a commit once I have perhaps all of the `5+` cases (appearing at least five times in Frank's list) done.

**Comment by josephwright** [comment]
The commit addresses all of the cases which occur at least five times, plus some that go with those in consistency terms, with the following exceptions. First, these are all expandable functions which therefore should normally not be used with `x`-type expansion:

- `\prop_item:Nx`
-  \str_if_eq:xxF`
- `\tl_if_empty:xTF`

There is also `\quark_if_no_value:VTF`, which I've skipped as I think really we should encourage use of `\quark_if_no_value:NTF` for this one: I wonder if the docs need adjustment to make that clearer?

**Comment by josephwright** [comment]
I've now covered the '3+' set too, but here have more queries. First, the list of those `x`-type skipped as the base is expandable and an `e`-type is (now) available:

- `\str_case:xn`
- `\str_if_eq:xnF`
- `\str_if_eq:xnT`
- `\tl_if_blank:xTF`

There are then a couple of cases of a non-expandable function which have variants that I think are not needed:

 - `\prop_put:cnf` could use `\prop_put:cnx`
- `\tl_put_right:Ne` could use `\tl_put_right:Nx` (unless `#` tokens are involved: doesn't look like it)

There is one `f`-type that I think could be covered by `e`-type: `\tl_if_blank:fTF`

----

There are also some cases that deserve more attention. First, all of the following likely can work without a variant: I think at least for the `V` type we have the question 'show we point to `set_eq`' and 'what is clearer'. (As noted earlier, there is an edge case where `V`-type expansion is useful, but I think that's out-of-scope for the kernel set.)

- `\int_set:NV`
- `\int_set:Nx`
- `\dim_set:NV`
- `\int_to_arabic:V`

Then we have two oddities. There is `\tl_if_eq:onTF`, which could be covered by the `V`-type or possibly by a `str` comparison. 	There is also `\use:nx`. We do have `\use:x`, but with `e`-type expansion available I do wonder if we should deprecate that. But on `\use:nx`, assuming we might favour `\use:ne` do we want some form of 'full-ish' set?

**Comment by car222222** [comment]
> There is one f-type that I think could be covered by e-type: `\tl_if_blank:fTF`"

Are there also examples of f-type that must not be changed to e-type?

**Comment by josephwright** [comment]
> > There is one f-type that I think could be covered by e-type: `\tl_if_blank:fTF`"
> 
> Are there also examples of f-type that must not be changed to e-type?

I think that is possible: if you have some user input that you want to expand but can't use `\protected@def` as you are in an expansion context, yu know that `f`-type expansion will stop rather than explode. But these are real edge cases.

**Comment by car222222** [comment]
I am not sure that I understand fully the discussion of the `\use:...` cases but I see no a priori reason for not deprecating all the x-type cases here, in favour of e-type (I recall many decades ago being in favour if driving an E-type:-). 

I can imagine reasons for wanting only "one-level expansion": does o-type give this expandably?

**Comment by josephwright** [comment]
What is interesting here, at least as a thought experiment, is the `e` vs `x` split. If one was starting today, with `\expanded` available, I suspect we wouldn't have `x`-type at all, but rather just have `e`-type and then all variant types do not affect expandability. So once _could_ argue that we should not add `x`-type variants at all, but rather add the matching `e` types. Or one could argue that as we have `x`-type, we have to live with it but should also match with `e`-type. You could even go as far as deprecating `x`-type - would be radical but would mean that there would be a simple rule - the current situation is a little tricky.

Just to show this does work:
```latex
\ExplSyntaxOn
\cs_generate_variant:Nn \cs_gset:Npn { Npe }
\cs_gset:Npn \fooa #1 { stuff~with~#1~and~##1 }
\cs_gset:Npx \foob #1 { stuff~with~#1~and~##1 }
\ifx\fooa\foob\TRUE\else\FALSE\fi
```

This might all be a step too far, but I'd welcome thoughts: @FrankMittelbach will obviously have some, would also welcome those from others!

**Comment by josephwright** [comment]
> I am not sure that I understand fully the discussion of the `\use:...` cases but I see no a priori reason for not deprecating all the x-type cases here, in favour of e-type (I recall many decades ago being in favour if driving an E-type:-).

We were just addressing the same thing :) Seems we agree on this area - I'll see what others see.

> I can imagine reasons for wanting only "one-level expansion": does o-type give this expandably?

`\foo:o{<tokens>}` is just `\expandafter\foo:n\expandafter{<tokens>}`, is that what you mean?

**Comment by josephwright** [comment]
I've thought a bit more about the `e`/`f`/`x` situation, checked a few things (e.g. usage in `siunitx`), etc. I suggest the following approach

- We do not add any new `x` type variants: that includes removing any I included in commits https://github.com/latex3/latex3/commit/b28ff08d5716eaf8fb1e079df57a94f6df1299dd and https://github.com/latex3/latex3/commit/695435731d8fff67e83b64ec174078318fabf953
- We switch from `x` to `e` for our code and for documented variants, and collect existing `x`-type ones _that are used in TeX Live_ into one place
- We promote/encourage moving from `x` to `e` and as the variants are removed from TL, we 'thin out' the list of retained variants: whilst we are committed to not removing base functions, I think we can remove variants under these circumstances
- We take a similar tack for `f`-type variants, except that we document that there are some limited use cases where `f`-type may be needed in place of `e`-type
- We thin out `o`-type variants to cover only those in current use in TeX Live (e.g. don't aim for consistent `o`-type coverage)

This would result in quite a clean up: we are left with the base types, the common variants `c`, `V`, `v`, e`, the 'limited use' variant `o`, and the 'extremely limited use' variant `f`, with `x` 'disappearing'. If we then assume that `o` and `f` are a special case (only to be provided if needed), we have only `N`, `c`, `n`, V`, `n`, `e` for 'consistent variant sets', and at least for functions with one or two arguments we can imagine providing full coverage. (Question about `\int_set:Nn` vs `\int_set:Ne`, etc., remains of course.)

**Comment by josephwright** [comment]
(Obviously not taking any action without discussion first!)

**Comment by josephwright** [comment]
Performance might be an issue, at least for `\tl_set:Nx`-equivalent, although we might avoid most of that hit within the kernel by using `:D` functions where appropriate, and beyond the kernel level we *might* feel the consistency is worth the cost. @FrankMittelbach  doubtless will have something to say.

**Comment by josephwright** [comment]
For example:
```latex
\RequirePackage{l3benchmark}
\ExplSyntaxOn
\cs_new:Npn \test_set:Nx #1#2
  {
    \tex_edef:D #1 {#2}
  }
\cs_new:Npn \test_set:Ne #1#2
  {
    \tex_edef:D #1
      { \tex_unexpanded:D \tex_expandafter:D { \tex_expanded:D {#2} } }
  }
\benchmark:n { \prg_replicate:nn { 1000 } { \test_set:Nx \foo { abc } } }
\benchmark:n { \prg_replicate:nn { 1000 } { \test_set:Ne \foo { abc } } }
```

**Comment by FrankMittelbach** [comment]
You don't show the figures @josephwright, but as far as I can see this means a 30% increase in time and in my opinion that is not acceptable, not even if we avoid it on the kernel level by using `:D` functions there after all, there are already quite time consuming packages in expl3 out there and if we slow them down considerably by asking them to use e when x does work for them then I wouldn't be surprised if that shows up as a noticeable slowdown. And if then those start also using `:D` to avoid that, or we use `:D` all over the place and also in packages then I don't think this can be considered a a move to consistency.

So my feeling is that we we shouldn't market "x" as something that should or need to be avoided generally but as something that has be avoided if you need an exandable command. But in many cases expandability is not really necessary.

Some more data from TL:
```
find . -type f -exec grep _set:Nx {} \; |wc -l      2618
find . -type f -exec grep -l _set:Nx {} \; |wc -l   245
```

**Comment by FrankMittelbach** [comment]
The f to e type move is a bit different and we probably should encourage to people to use and replace f with e normally (might even be faster?). But at the same time would should explain relatively thoroughly in which circumstances f should be used instead of e.

**Comment by FrankMittelbach** [comment]
> (Question about |\int_set:Nn| vs |\int_set:Ne|, etc., remains of course.)

isn't that something that could and perhaps should be covered by simply 
defining

`\cs_new_eq:NN \int_set:Ne \int_set:Nn`

because the primitive really does "e" here? Then this allows the use of 
e without an unnecessary slowdown (at the cost of one csname). Or we 
could leave out the "e" and tell people because the primitive does 
expansion one should unnecessary define and use such a variant as it 
only slows down the processing.

personally I think the latter is better even if it means that some of 
TeX engine behavior is showing through --- but this is the case for a 
long time now.

(there is obviously no way to make the Nn not expand at this stage, as 
it would be a really bad breaking change for no good reason and would 
most certainly fall over in many places even if we tried to adjust all 
code that we know about).

**Comment by josephwright** [comment]
> The f to e type move is a bit different and we probably should encourage to people to use and replace f with e normally (might even be faster?). But at the same time would should explain relatively thoroughly in which circumstances f should be used instead of e.

I'll look at this next, then: I think it's quite uncontroversial.

**Comment by josephwright** [comment]
> You don't show the figures @josephwright, but as far as I can see this means a 30% increase in time

I suspect this would be a tiny impact 'in the wild', and I suspect becomes must less significant when you look at something higher-level, e.g. `\prop_put:Nnx` vs `\prop_put:Nne`. However, I'm not going to push it - I just wanted to raise the question as it came to my mind.

**Comment by FrankMittelbach** [comment]
I agree this is difficult to just just by looking at the number of calls. Perhaps we would get a bit more reliable data if we try to run the test suite with a format that sets various `Nx` commands to `Ne` near the end of the format and then check if this shows any noticable difference? Obviously also only a limited snapshot but if that already shows a significant difference we would know at least that this *is* an issue.

**Comment by josephwright** [comment]
@FrankMittelbach I'll see what I can do: perhaps in the first instance with a branch that alters `\tl_set:Nx`, etc., and then I'll benchmark. Results will be useful so we have a record I suspect.

**Comment by wspr** [comment]
This might be a silly idea, but if you want to go this way what about redefining x to mean e, then adding a new variant for the old x? This would have the benefit of having old code keep working, people don’t need to switch to a new letter, with the faster unexpandable approach still available if needed for performance

**Comment by davidcarlisle** [comment]
it's tempting but wouldn't keep all code working as `##` would make `##`  rather than `#`

**Comment by FrankMittelbach** [comment]
don't think that works @wspr. a) it would make all code slower because right now x is used nearly everywhere and b) the "x" is firmly associated for 2.5 decades now with with `\xdef` aka making something unexpandable

**Comment by car222222** [comment]
Were it not for @davidcarlisle pointing out (roughly speaking:-) the `##` problem, I would support @wspr in his idea.  

The problem is that most coders do not understand much about `#`-doubling, such as when it does or does not happen or when it needs to be done.

**Comment by car222222** [comment]
Some questions concerning the claimed 30% slow down:

Do we have any theoretical reasoning for this large difference? 
Exactly what takes 30% longer?  
Is it because the primitive \expanded is that much slower than the use of \edef? 
Or is it because of the \expandafter(s) that is/are needed?
Or is there some other reason?

**Comment by FrankMittelbach** [comment]
> Some questions concerning the claimed 30% slow down:

it is not "claimed" it is the slowdown that is seen in @josephwright's test file above.

> Do we have any theoretical reasoning for this large difference? Exactly what takes 30% longer? Is it because the primitive \expanded is that much slower than the use of \edef? Or is it because of the \expandafter(s) that is/are needed? Or is there some other reason?

I think it islargely because `#2` is processed/read twice instead of once so with longer arguments it might even get worse.

**Comment by FrankMittelbach** [comment]
> it's tempting but wouldn't keep all code working as `##` would make `##` rather than `#`

which makes it a no-go imho and should have been point a) in my list of why not

**Comment by car222222** [comment]
> it is not "claimed" 
I thought it was mentioned earlier?

> it is the slowdown that is seen in @josephwright's test file above.

No doubt, but he never told us the results!

> I think it is largely because #2 is processed/read twice instead of once, so with longer arguments it might even get worse.

That sounds reasonable.

**Comment by car222222** [comment]
So `\unexpanded` "processes the whole content" rather than just "adjusting some links"?

**Comment by FrankMittelbach** [comment]
> So `\unexpanded` "processes the whole content" rather than just "adjusting some links"?

probably not and quite likely my initial guess was wrong and it has to do more with the number of primitives being used or not used (I remember in the good old days taking the number of lines in a `\tracingall` log was a reasonable measure for the running time). Dunno; the difference is there, and how much it matters would be something one can only see in larger real life documents.

**Comment by josephwright** [comment]
I think I've found a nice clear-cut reason we need `x`-type, independent of performance. In various places we use the `\iffalse{\fi}` trick to 'escape' an `\edef`. That fails with `\expanded`, which would affect a number of core functions. That's a clear blocker, so I think we can conclude that `x`-type stays, but with the expandable/non-expandable split.

(If I ever try another post-`expl3` experiment, I'll of course pick up on this.)

**Comment by josephwright** [comment]
I'm moving on into the lower parts of the list, but am now planning to work on a per-module basis and review a bit more as I go. I'm starting with `prop`, where a search of TL has let me clean up a few variants and standardise more: there are a _lot_ of different combos for `\prop_put:Nnn` in particular.

Looking beyond that, in `clist` I'm wondering about `\clist_map_inline:Vn`, as the space handling means it might be slightly different to `\clist_map_inline:Nn` if applied to a `tl`. But, as this is not expandable, one could simply assign to a `tl` then use the `N`-type function, so I'm torn.

That and  similar edge cases I'll skip over and try to get a decent amount more standardisation done. Commit comments will hopefully make the decisions clear.

**Comment by josephwright** [comment]
I'm going to take another look at `x`/`e`: @Skillmon made some suggestions I'd like to explore

**Comment by lnzspraj** [comment]
> > So `\unexpanded` "processes the whole content" rather than just "adjusting some links"?
> 
> probably not and quite likely my initial guess was wrong and it has to do more with the number of primitives being used or not used (I remember in the good old days taking the number of lines in a `\tracingall` log was a reasonable measure for the running time). Dunno; the difference is there, and how much it matters would be something one can only see in larger real life documents.

It does sort-of process it, it has to scan for balanced braces, so it steps through the list at least once. But the offset is something like go from O(n) to O(2n), so the effect should be linear. Benchmarking the code

```tex
\RequirePackage{l3benchmark}
\ExplSyntaxOn
\cs_new_protected:Npn \test_set:Nx #1#2
  {
    \tex_edef:D #1 {#2}
  }
\cs_new_protected:Npn \test_set:Ne #1#2
  {
    \tex_edef:D #1
      { \tex_unexpanded:D \tex_expandafter:D { \tex_expanded:D {#2} } }
  }
\int_step_inline:nnn {0} {4} {
\typeout{\fp_eval:n{10^#1}}
\exp_args:Ne \benchmark:n { \test_set:Nx \noexpand\foo { \prg_replicate:nn { \fp_eval:n { 10^#1 } } { abc } } }
\exp_args:Ne \benchmark:n { \test_set:Ne \noexpand\foo { \prg_replicate:nn { \fp_eval:n { 10^#1 } } { abc } } }
}
\stop
```

I get:

```none
1
1.99e-7 seconds (0.873 ops)
3.3e-7 seconds (1.46 ops)
10
7.49e-7 seconds (3.3 ops)
1.14e-6 seconds (5.08 ops)
100
6.33e-6 seconds (27.9 ops)
9.46e-6 seconds (41.7 ops)
1000
6.25e-5 seconds (276 ops)
9.52e-5 seconds (420 ops)
10000
6.24e-4 seconds (2.76e3 ops)
9.29e-4 seconds (4.11e3 ops)
```

**Comment by car222222** [comment]
"go from O(n) to O(2n)" where, I guess, n is the length of the argument?

**Comment by Skillmon** [comment]
@car222222 yes, as you can see from the timings, a ten times as long token list takes roughly ten times as long to store, regardless which version is used, the ratio between the two stays constant. (and sorry, the comment above was posted with the wrong account).

**Comment by car222222** [comment]
Sure, I could see, just checking:-).

So it was you all,the time:-).

**Comment by josephwright** [comment]
I think we are done here for the present: I might look at another pass once the `e`-type changes have propagated.
